lib/iris/util.py

```diff
@@ -1593,6 +1593,173 @@ def demote_dim_coord_to_aux_coord(cube, name_or_coord):
+                  y                           x
+                  latitude                    x
+                  longitude                   x
+    """
+    if isinstance(name_or_coord, six.string_types):
+        coord = cube.coord(name_or_coord)
+    elif isinstance(name_or_coord, iris.coords.Coord):
+        coord = name_or_coord
+    else:
+        # Don't know how to handle this type
+        msg = ("Don't know how to handle coordinate of type {}. "
+               "Ensure all coordinates are of type six.string_types or "
+               "iris.coords.Coord.")
+        msg = msg.format(type(name_or_coord))
+        raise TypeError(msg)
+
+    coord_dims = cube.coord_dims(coord)
+
+    if len(coord_dims) == 1:
+        # Do nothing
+        return cube
+    elif len(coord_dims) > 2:
+        # Don't currently support coords with more than 2 dimensions
+        msg = ("Collapsing coordinates with more than 2 dimensions "
+               "is not currently supported. Tried to collapse {} with"
+               " {} dimensions. ")
+        msg = msg.format(coord.name(), coord_dims)
+        raise NotImplementedError(msg)
+
+    return flatten_cube(cube, coord_dims)
+
+
+def flatten_cube(cube, dims=None):
+    """
+    Flatten the cube along the specified dimensions or all (by default).
+    Coordinates along those dimensions are flattened. DimCoords are
+    demoted to AuxCoords before flattening. The new (anonymous)
+    flattened dimension will be along the first dimension of the
+    specified dims.
+
+    * cube
+        An instance of :class:`iris.cube.Cube`
+
+    * dims (:class:`list`, :class:`tuple` etc.):
+        The dimensions of the cube to flatten
+
+    For example::
+
+        >>> print(cube)
+        rainfall_rate / (kg m-2 s-1)        (x: 5; y: 3)
+             Dimension coordinates:
+                  x                           x     -
+                  y                           -     x
+             Auxiliary coordinates:
+                  latitude                    x     x
+                  longitude                   x     x
+
+        >>> flatten_cube(cube)
+        >>> print(cube)
+        rainfall_rate / (kg m-2 s-1)        (--: 15)
+             Auxiliary coordinates:
+                  x                           x
+                  y                           x
+                  latitude                    x
+                  longitude                   x
+    """
+    import numpy as np
+    from iris.cube import Cube
+    from iris.coords import AuxCoord
+
+    all_dims = range(cube.ndim)
+
+    dims = dims if dims is not None else all_dims
+
+    other_dims = list(set(all_dims) - set(dims))
+    shape_array = np.asarray(cube.shape)
+    # The new (flat) dimension will be in the first dimension of the aux coord
+    flat_dim = dims[0]
+    dim_shape = shape_array[list(dims)]
+    new_dim_shape = np.product(dim_shape)
+    new_shape = np.insert(shape_array[other_dims], flat_dim, new_dim_shape)
```

Comment:
There are problems with this calc, even when the dims are in order.
Examples testing this part of the code, extracted into a standalone function ...
```
def newshape(shape, dims):
   all_dims = range(len(shape))
   shape_array = np.asarray(shape)
   flat_dim = dims[0]
   dim_shape = shape_array[list(dims)]
   new_dim_shape = np.product(dim_shape)
   new_shape = np.insert(shape_array[other_dims], flat_dim, new_dim_shape)
   return new_shape
```
Then
`newshape((10, 11, 12, 13), (0, 2)) --> [120,  11,  13]` is ok, but ..
`newshape((10, 11, 12, 13), (2, 0)) --> [ 11,  13, 120]` is wrong : (reverse dims order) : should be same as previous (??)
Likewise ...
`newshape((10, 11, 12, 13), (1, 2)) --> [ 11, 132,  13]` is wrong : should give  (10, 132, 12)